### PR TITLE
[FEAT] Grazing Range Contract

### PR DESCRIPTION
**Pull request by abraham-alpaca**
## Description
This is a feature PR for Grazing Range Contract.

## Why?
Because we want our beloved alpacas to have more choices on staking, right now staking IBAlpaca can gain many benefits from this contract.

## ChangeLogs:
- Implement Grazing Range Contract.
- Create Unit tests for grazing range contract based on test cases.

### CODE REVIEW

File: contracts/6/token/GrazingRange.sol [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        }
+        // @dev when couldn't find any reward info, it means that timestamp exceed endblock
+        // so return the latest reward info
+        return rewardInfo[len-1].rewardPerBlock;
+    }
+
+
+    // @notice Return reward multiplier over the given _from to _to block.
+    function getMultiplier(uint256 _from, uint256 _to, uint256 _endBlock) public pure returns (uint256) {
+        if (_to <= _endBlock) {
```
Comment (almost-llamas) [inline contracts/6/token/GrazingRange.sol:140]: should we concern the case where _from is greater than _to?

Comment (abraham-alpaca) [reply contracts/6/token/GrazingRange.sol:140]: can validate, but it's not that severed to validate these params.

File: contracts/6/token/GrazingRange.sol [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        RewardInfo[] storage rewardInfo = campaignRewardInfo[_campaignID];
+        uint256 accRewardPerShare = campaign.accRewardPerShare;
+        uint256 totalSupply = campaign.totalStaked;
+        if (block.number > campaign.lastRewardBlock && totalSupply != 0) {
+            uint256 rewardInfoLen = rewardInfo.length;
+            uint256 multiplier;
+            uint256 rewardPerBlock;
+            for (uint256 i = 0; i < rewardInfoLen; ++i) {
+                RewardInfo memory info = rewardInfo[i];
+                rewardPerBlock = info.rewardPerBlock;
```
Comment (almost-llamas) [inline contracts/6/token/GrazingRange.sol:162]: no need to declare this

File: contracts/6/token/GrazingRange.sol (hunk 1 above)
Comment (abraham-alpaca) [reply contracts/6/token/GrazingRange.sol:140]: well, actually this function shouldn't have any validation, just get multiplier from input, that's all

File: contracts/6/token/GrazingRange.sol [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    // @notice Withdraw Staking tokens from STAKING.
+    function withdraw(uint256 _campaignID, uint256 _amount) external nonReentrant {
+        _withdraw(_campaignID, _amount);
+    }
+
+    // @notice internal method for withdraw (withdraw and harvest method depend on this method)
+    function _withdraw(uint256 _campaignID, uint256 _amount) internal {
+        CampaignInfo storage campaign = campaignInfo[_campaignID];
+        UserInfo storage user = userInfo[_campaignID][msg.sender];
+        require(user.amount >= _amount, "withdraw: withdraw amount exceed available amount");
```
Comment (spicysquid168) [inline contracts/6/token/GrazingRange.sol:247]: Use pattern above for revert msg

**Review by spicysquid168 — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Pull request merged**